cpp/src/arrow/io/mremap.h

```diff
@@ -0,0 +1,86 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+#ifdef _WIN32
+#include "arrow/io/mman.h"
+#undef Realloc
+#undef Free
+#else
+#include <sys/mman.h>
+#include <unistd.h>
+#endif
+
+namespace arrow {
+namespace internal {
+// define a wrapper around mremap which the same arguments on linux and on windows
+// should only be called with writable files
+static void* MemoryMapRemap(void* addr, size_t old_size, size_t new_size, int fildes) {
```

Comment:
Rather than making it a static inline function, just make it a regular function that is defined in the `.cc` file, IMHO. We won't benefit from inlining here.